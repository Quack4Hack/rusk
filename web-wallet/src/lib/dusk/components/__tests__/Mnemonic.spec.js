import {
	afterEach,
	describe,
	expect,
	it
} from "vitest";
import {
	cleanup,
	fireEvent,
	render
} from "@testing-library/svelte";
import { Mnemonic } from "..";

/** @type {string[]} */
const enteredSeed = [];

/** @type {string[]} */
const seed = ["auction", "tribe", "type", "torch", "domain", "auction",
	"lyrics", "mouse", "alert", "fabric", "snake", "ticket"];

describe("Mnemonic", () => {
	afterEach(cleanup);

	it("should render the \"Mnemonic\" component in the authenticate state", async () => {
		const { container } = render(Mnemonic, {
			props: {
				enteredMnemonicPhrase: enteredSeed,
				type: "authenticate"
			}
		});

		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render the \"Mnemonic\" component in the validate state", () => {
		const { container } = render(Mnemonic, {
			props: {
				mnemonicPhrase: seed,
				type: "validate"
			}
		});

		expect(container.firstChild).toMatchSnapshot();
	});

	it("should display all the words in the order they have been clicked", async () => {
		const { container, getAllByRole } = render(Mnemonic, {
			props: {
				enteredMnemonicPhrase: enteredSeed,
				mnemonicPhrase: seed,
				type: "validate"
			}
		});

		const buttons = container.querySelectorAll("button[data-value]");

		for (const word of buttons) {
			await fireEvent.click(word);
			expect(word).toBeDisabled();
		}

		const enteredPhrase = getAllByRole("listitem");

		enteredPhrase.forEach((word, index) => {
			expect(word.textContent).toBe(buttons[index].textContent);
		});
	});

	it("should revert the most recent word on Undo click", async () => {
		const { container, getByText, getAllByRole } = render(Mnemonic, {
			props: {
				enteredMnemonicPhrase: enteredSeed,
				mnemonicPhrase: seed,
				type: "validate"
			}
		});

		const buttons = container.querySelectorAll("button[data-value]");

		// Enters the first 5 words
		for (let i = 0; i < 5; i++) {
			await fireEvent.click(buttons[i]);
			expect(buttons[i]).toBeDisabled();
		}

		const undoButton = getByText("Undo");

		// Presses the "Undo button"
		await fireEvent.click(undoButton);

		const enteredPhrase = getAllByRole("listitem");

		// Verify that the content of each list item matches the corresponding button's text content
		// Loop until the 4th item (index 3)
		for (let i = 0; i <= 3; i++) {
			expect(enteredPhrase[i].textContent).toBe(buttons[i].textContent);
		}

		// The 5th item (index 4) and next should be empty
		for (let i = 4; i <= 11; i++) {
			expect(enteredPhrase[i].textContent).toBe("_____");
		}
	});
});
