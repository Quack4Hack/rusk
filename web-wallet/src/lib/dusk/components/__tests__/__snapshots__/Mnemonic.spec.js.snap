// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mnemonic > should render the "Mnemonic" component in the authenticate state 1`] = `
<div
  class="dusk-mnemonic"
>
  <div
    class="dusk-mnemonic__actions-wrapper"
  >
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
      disabled=""
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M10.5,7A6.5,6.5 0 0,0 4,13.5A6.5,6.5 0 0,0 10.5,20H14V18H10.5C8,18 6,16 6,13.5C6,11 8,9 10.5,9H16.17L13.09,12.09L14.5,13.5L20,8L14.5,2.5L13.08,3.91L16.17,7H10.5M18,18H16V20H18V18Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Undo
      </span>
      
    </button>
  </div>
   
  <ol
    class="dusk-words"
  >
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
  </ol>
   
  <div
    class="dusk-mnemonic__authenticate-actions-wrapper"
  >
    <input
      class="dusk-textbox dusk-textbox-text"
      maxlength="8"
      placeholder="Enter word 1"
      type="text"
    />
    
     
    
  </div>
</div>
`;

exports[`Mnemonic > should render the "Mnemonic" component in the validate state 1`] = `
<div
  class="dusk-mnemonic"
>
  <div
    class="dusk-mnemonic__actions-wrapper"
  >
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
      disabled=""
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M10.5,7A6.5,6.5 0 0,0 4,13.5A6.5,6.5 0 0,0 10.5,20H14V18H10.5C8,18 6,16 6,13.5C6,11 8,9 10.5,9H16.17L13.09,12.09L14.5,13.5L20,8L14.5,2.5L13.08,3.91L16.17,7H10.5M18,18H16V20H18V18Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Undo
      </span>
      
    </button>
  </div>
   
  <ol
    class="dusk-words"
  >
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
    <li
      class="dusk-words__word"
    >
      _____
    </li>
  </ol>
   
  <div
    class="dusk-mnemonic__validate-actions-wrapper"
  >
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="auction"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        auction
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="tribe"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        tribe
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="type"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        type
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="torch"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        torch
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="domain"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        domain
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="auction"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        auction
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="lyrics"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        lyrics
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="mouse"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        mouse
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="alert"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        alert
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="fabric"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        fabric
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="snake"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        snake
      </span>
    </button>
    
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default"
      data-value="ticket"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        ticket
      </span>
    </button>
    
  </div>
</div>
`;
